horovod/run/elastic/worker.py

```diff
@@ -0,0 +1,116 @@
+# Copyright 2020 Uber Technologies, Inc. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+
+from __future__ import absolute_import
+
+import os
+import threading
+
+from horovod.run.common.util import network, secret
+from horovod.run.elastic.rendezvous import PUT_WORKER_ADDRESSES
+from horovod.run.http.http_client import put_data_into_kvstore
+
+
+HOROVOD_GLOO_RENDEZVOUS_ADDR = 'HOROVOD_GLOO_RENDEZVOUS_ADDR'
+HOROVOD_GLOO_RENDEZVOUS_PORT = 'HOROVOD_GLOO_RENDEZVOUS_PORT'
+HOROVOD_GLOO_IFACE = 'HOROVOD_GLOO_IFACE'
+HOROVOD_HOSTNAME = 'HOROVOD_HOSTNAME'
+HOROVOD_LOCAL_RANK = 'HOROVOD_LOCAL_RANK'
+
+
+class HostsUpdatedRequest(object):
+    """Notifies worker that the set of available hosts/slots has changed."""
+    def __init__(self, timestamp):
+        self.timestamp = timestamp
+
+
+class HostsUpdatedResponse(object):
+    pass
+
+
+class WorkerNotificationManager(object):
+    def __init__(self):
+        self._lock = threading.Lock()
+        self._service = None
+        self._listeners = set()
+
+    def init(self, rendezvous_addr=None, rendezvous_port=None,
+             nic=None, hostname=None, local_rank=None):
+        with self._lock:
+            if self._service:
+                return
+
+            rendezvous_addr = rendezvous_addr or os.environ.get(HOROVOD_GLOO_RENDEZVOUS_ADDR)
+            if not rendezvous_addr:
+                return
+
+            rendezvous_port = rendezvous_port if rendezvous_port is not None else \
```

Comment:
`rendezvous_port or int(os.environ.get(HOROVOD_GLOO_RENDEZVOUS_PORT))`

Reply:
Same as local_rank.

Reply:
`rendezvous_port` can be `None` here when this env var is not set (which should be set by the launching code), but exception will be something like `could not connect to http://127.0.0.1:None/worker_addresses/hostname:local_rank`, so a meaningful exception that this env var is missing would improve the situation